pkg/provisioner/ironic/raid.go

```diff
@@ -0,0 +1,172 @@
+package ironic
+
+import (
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic/devicehints"
+
+	"github.com/pkg/errors"
+)
+
+// setTargetRAIDCfg set target for RAID configuration steps
+func setTargetRAIDCfg(p *ironicProvisioner, ironicNode *nodes.Node) (err error) {
+	var logicalDisks []nodes.LogicalDisk
+
+	// Build target for RAID configuration steps
+	logicalDisks, err = buildTargetRAIDCfg(p.host.Status.Provisioning.RAID)
+	if len(logicalDisks) == 0 || err != nil {
+		return
+	}
+
+	// set root volume
+	if p.host.Status.Provisioning.RootDeviceHints == nil {
+		logicalDisks[0].IsRootVolume = new(bool)
+		*logicalDisks[0].IsRootVolume = true
+	} else {
+		p.log.Info("rootDeviceHints is used, the first volume of raid will not be set to root")
+	}
+
+	// Set target for RAID configuration steps
+	return nodes.SetRAIDConfig(
+		p.client,
+		ironicNode.UUID,
+		nodes.RAIDConfigOpts{LogicalDisks: logicalDisks},
+	).ExtractErr()
+}
+
+// A private method to build RAID disks
+func buildTargetRAIDCfg(raid *metal3v1alpha1.RAIDConfig) (logicalDisks []nodes.LogicalDisk, err error) {
+	if raid == nil {
+		return
+	}
+
+	// build logicalDisks
+	if len(raid.HardwareRAIDVolumes) != 0 {
+		logicalDisks, err = buildTargetHardwareRAIDCfg(raid.HardwareRAIDVolumes)
+	} else if len(raid.SoftwareRAIDVolumes) != 0 {
+		logicalDisks, err = buildTargetSoftwareRAIDCfg(raid.SoftwareRAIDVolumes)
+	}
+
+	return
+}
+
+// A private method to build hardware RAID disks
+func buildTargetHardwareRAIDCfg(volumes []metal3v1alpha1.HardwareRAIDVolume) (logicalDisks []nodes.LogicalDisk, err error) {
+	var (
+		logicalDisk    nodes.LogicalDisk
+		nameCheckFlags map[string]struct{} = make(map[string]struct{})
+	)
+
+	if len(volumes) == 0 {
+		return
+	}
+
+	for _, volume := range volumes {
+		// Check volume's name
+		if volume.Name != "" {
+			_, exist := nameCheckFlags[volume.Name]
+			if exist {
+				return nil, errors.Errorf("name of volume must be unique")
+			}
+			nameCheckFlags[volume.Name] = struct{}{}
+		}
+		// Build logicalDisk
+		logicalDisk = nodes.LogicalDisk{
+			SizeGB:     volume.SizeGibibytes,
+			RAIDLevel:  nodes.RAIDLevel(volume.Level),
```

Comment:
Is there any chance that this type conversion could fail? The code would panic if it does, right?

Reply:
The type of  `volume.Level` is string,  the definition of `nodes.RAIDLevel` is `type RAIDLevel string`. They are essentially the same, so I don’t think this type conversion will fail.